src/net/socket.rs

```diff
@@ -33,47 +30,64 @@ impl Socket {
+        let socket = UdpSocket::bind(addresses)?;
+        socket.set_nonblocking(true)?;
+        let (event_sender, event_receiver) = unbounded();
+        let (packet_sender, packet_receiver) = unbounded();
+        Ok((
+            Socket {
+                socket,
+                config,
+                connections: ActiveConnections::new(),
+                recv_buffer: Vec::new(),
+                link_conditioner: None,
+                event_sender,
+                packet_receiver,
+            },
+            packet_sender,
+            event_receiver,
+        ))
     }
 
     /// Entry point to the run loop. This should run in a spawned thread since calls to `poll.poll`
     /// are blocking.
     pub fn start_polling(&mut self) -> Result<()> {
-        let poll = Poll::new()?;
-
-        poll.register(self, SOCKET, Ready::readable(), PollOpt::edge())?;
-
-        let mut events = Events::with_capacity(self.config.socket_event_buffer_size);
-        let packet_receiver = self.packet_receiver.clone();
         // Nothing should break out of this loop!
+        let mut to_send: Vec<Packet>;
         loop {
-            self.handle_idle_clients();
-            if let Err(e) = poll.poll(&mut events, self.config.socket_polling_timeout) {
-                error!("Error polling the socket: {:?}", e);
-            }
-            if let Err(e) = self.process_events(&mut events) {
-                error!("Error processing events: {:?}", e);
-            }
-
-            for packet in packet_receiver.try_iter() {
-                // XXX: I'm fairly certain this isn't exactly safe. I'll likely need to add some
-                // handling for when the socket is blocked on send. Worth some more research.
-                // Alternatively, I'm sure the Tokio single threaded runtime does handle this for us
-                // so maybe it's work switching to that while providing the same interface?
-                if let Err(e) = self.send_to(packet) {
-                    error!("Error sending packet: {:?}", e);
+            // First we pull any newly arrived packets and handle them
+            match self.recv_from() {
+                Ok(result) => match result {
+                    Some(packet) => {
+                        match self.event_sender.send(SocketEvent::Packet(packet)) {
+                            Ok(_) => {}
+                            Err(e) => {
+                                error!("Error sending SocketEvent: {:?}", e);
+                            }
+                        };
+                    }
+                    None => {
+                        error!("Empty packet received");
+                    }
+                },
+                Err(e) => {
+                    error!("Error receiving packet: {:?}", e);
+                }
+            };
+
+            // Now grab all the packets waiting to be sent and send them
+            to_send = self.packet_receiver.try_iter().collect();
```

Comment:
This is still doing an allocation per loop iteration. Basically, any time you call `collect()` you're making a new `Vec`